### PR TITLE
<deque>: Use _Next_iter and _Prev_iter

### PR DESCRIPTION
**Pull request by CaseyCarter**
Fixes DevCom-1134328 / [AB#1166791](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1166791).

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
I wonder if we should have a more comprehensive test (like our `Evil` class with overloaded `operator,` and `operator&`), given that this code seems to be conformant and there could be issues elsewhere in the STL especially during future refactoring, but certainly this is a strict improvement over the status quo. Perhaps we should file a `test` issue to further enhance coverage.

**Comment by BillyONeal** [comment]
@StephanTLavavej The difference is that unlike the `&`/`,` situation, there's nothing we can really do to guard against this kind of `ADL-injected-ill-formed` problem in the general case. For example, allowing this in general would prohibit calling any standard library algorithms from any of the containers, because the algorithms have expressions like `iter+42` inside which are vulnerable in the same way this is vulnerable. Even if the spec doesn't require this right now I think we should fix the spec to prohibit and behave as if the spec does prohibit such ADL hijacking of operators when neither of the operands are a program-defined type (here, `_Deque_iterator` and `int`).

**Comment by CaseyCarter** [comment]
Formally `iter+42` is allowed only if `iter` is a random-access iterator whose difference type is `int`. We guard against this in the general case by only adding or subtracting values of an iterator's difference type (ignoring cv-qualifiers) to an iterator. We've been very careful about the types of variables that we add and subtract from iterators, but not so much with literals.

**Comment by BillyONeal** [comment]
> Formally `iter+42` is allowed only if `iter` is a random-access iterator whose difference type is `int`.

Would that have protected against this particular case?

**Comment by CaseyCarter** [comment]
> > Formally `iter+42` is allowed only if `iter` is a random-access iterator whose difference type is `int`.
> 
> Would that have protected against this particular case?

Yes. The syndrome here is that we have two functions:
```c++
struct S {
  template <class T> S(T&&); // Everything converts to S
};

void f(S, int);              // Unintentional ADL-hijacking function
void f(iterator, ptrdiff_t); // Actual function we want to call
```
and we're evaluating `f(iterator{}, 1)` with an iterator whose difference type is `ptrdiff_t`. This works as we intend in 32-bit compiles - where `ptrdiff_t` is `int` - but not in 64-bit compiles when it's `long long`. Our `f` has a better ICS (implicit conversion sequence) for the first argument - exact match - but the user `f` has a better ICS for the second argument - again exact match. Overload resolution is ambiguous. `f(iterator{}, ptrdiff_t{1})` would unambiguously call the correct function.

File: stl/inc/deque [hunk 1]
```diff
@@ -1064,15 +1064,15 @@ public:
         _STL_VERIFY(!empty(), "back() called on empty deque");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return *(_Unchecked_end() - 1);
+        return *(_Prev_iter(_Unchecked_end()));
```
Comment (StephanTLavavej) [inline stl/inc/deque:1067]:
```suggestion
        return *_Prev_iter(_Unchecked_end());
```

File: stl/inc/deque [hunk 2]
```diff
@@ -1064,15 +1064,15 @@ public:
...
     }
 
     _NODISCARD const_reference back() const noexcept /* strengthened */ {
 #if _CONTAINER_DEBUG_LEVEL > 0
         _STL_VERIFY(!empty(), "back() called on empty deque");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return *(_Unchecked_end() - 1);
+        return *(_Prev_iter(_Unchecked_end()));
```
Comment (StephanTLavavej) [inline stl/inc/deque:1075]:
```suggestion
        return *_Prev_iter(_Unchecked_end());
```

**Comment by StephanTLavavej** [comment]
Thanks for investigating and fixing this surprising bug! I never would have thought that an STL container couldn't say `iter + 1` for its own iterators.

**Pull request merged**